File: python/optool/problem.py
from abc import ABC, abstractmethod

from numpy import array, ndarray

class Problem(ABC):
    def __init__(self, dimension: int, lowerBounds: ndarray, upperBounds: ndarray):
        self.__dimension = dimension
        
        if not isinstance(lowerBounds, (list, ndarray)):
            raise TypeError('lowerBounds: Expecting list or ndarray but got {}'.format(type(lowerBounds).__name__))
        if not isinstance(upperBounds, (list, ndarray)):
            raise TypeError('upperBounds: Expecting list or ndarray but got {}'.format(type(lowerBounds).__name__))
        if dimension < 1:
            raise ValueError('dimension: Must be greater then 0')
        if dimension != len(lowerBounds):
            raise ValueError('lowerBounds: Wrong dimension')
        if dimension != len(upperBounds):
            raise ValueError('upperBounds: Wrong dimension')
        
        self.__lowerBounds = array(lowerBounds)
        self.__upperBounds = array(upperBounds)
        self.__rangeBounds = self.__upperBounds - self.__lowerBounds
    
    @property
    def dimension(self):
        return self.__dimension
    
    @property
    def lowerBounds(self):
        return self.__lowerBounds

    @property
    def upperBounds(self):
        return self.__upperBounds
    
    @property
    def rangeBounds(self):
        return self.__rangeBounds
        
    @abstractmethod
    def evaluate(self, solutions: ndarray) -> ndarray:
        pass

File: java/joptool/src/main/java/com/gritti/joptool/mono/Problem.java
package com.gritti.joptool.mono;

import com.gritti.joptool.core.Matrix;

public abstract class Problem {
	protected Matrix lowerBounds;
	protected Matrix upperBounds;
	protected Matrix rangeBounds;

	protected int dimension;

	public Problem(int dimension, double[] lowerBounds, double[] upperBounds) {
		this.dimension = dimension;
		this.lowerBounds = Matrix.from(lowerBounds, 1);
		this.upperBounds = Matrix.from(upperBounds, 1);
		this.rangeBounds = Matrix.subtract(this.upperBounds, this.lowerBounds);
	}
}
